torchrl/envs/model_based/common.py

```diff
@@ -0,0 +1,150 @@
+                out_keys=["next_hidden_observation"],
+            ),
+            TensorDictModule(
+                nn.Linear(4, 1),
+                in_keys=["hidden_observation"],
+                out_keys=["reward"],
+            ),
+        )
+        world_model = MyWorldModel()
+        env = MyMBEnv(world_model)
+        td = env.reset()
+        env.rollout(td, max_steps=10)
+    ```
+
+
+    Properties:
+        - observation_spec (CompositeSpec): sampling spec of the observations;
+        - action_spec (TensorSpec): sampling spec of the actions;
+        - reward_spec (TensorSpec): sampling spec of the rewards;
+        - batch_size (torch.Size): number of environments contained in the instance;
+        - device (torch.device): device where the env input and output are expected to live
+        - is_done (torch.Tensor): boolean value(s) indicating if the environment has reached a done state since the
+            last reset
+
+    Args:
+        world_model (nn.Module): model that generates world states and its corresponding rewards;
+        device (torch.device): device where the env input and output are expected to live
+        dtype (torch.dtype): dtype of the env input and output
+        batch_size (torch.Size): number of environments contained in the instance
+
+    Methods:
+        step (TensorDict -> TensorDict): step in the environment
+        reset (TensorDict, optional -> TensorDict): reset the environment
+        set_seed (int -> int): sets the seed of the environment
+        rand_step (TensorDict, optional -> TensorDict): random step given the action spec
+        rollout (Callable, ... -> TensorDict): executes a rollout in the environment with the given policy (or random
+            steps if no policy is provided)
+
+    """
+
+    def __init__(
+        self,
+        world_model: TensorDictModule,
+        params: Optional[List[torch.Tensor]] = None,
+        buffers: Optional[List[torch.Tensor]] = None,
+        device: DEVICE_TYPING = "cpu",
+        dtype: Optional[Union[torch.dtype, np.dtype]] = None,
+        batch_size: Optional[torch.Size] = None,
+    ):
+        super(ModelBasedEnv, self).__init__(
+            device=device, dtype=dtype, batch_size=batch_size
+        )
+        self.world_model = world_model
+        self.world_model_params = params
+        self.world_model_buffers = buffers
+        self._inplace_update = False
+
+    def set_specs_from_env(self, env: EnvBase):
+        """
+        Sets the specs of the environment from the specs of the given environment.
+        """
+        self.observation_spec = deepcopy(env.observation_spec)
+        self.action_spec = deepcopy(env.action_spec)
+        self.reward_spec = deepcopy(env.reward_spec)
+        self.input_spec = deepcopy(env.input_spec)
+
+    def _step(
+        self,
+        tensordict: TensorDict,
+    ) -> TensorDict:
+        # step method requires to be immutable
+        tensordict_out = tensordict.clone()
```

Comment:
we can even clone with recursive=False, I think. It will be much faster and less mem-consuming